qemu/tests/slof_boot.py

```diff
@@ -0,0 +1,165 @@
+            boot options.
+            3.2 Check the boot order which should be just disk.
+        Scenario 4:
+            4.1 Boot a guest with an empty disk and nic, specify this device
+            bootindex=0, then set "order=cdn,once=n,menu=off,strict=off" for
+            boot options.
+            4.2 Check the boot order which should be disk->nic.
+    :param test: Qemu test object.
+    :param params: Dictionary with the test .
+    :param env: Dictionary with test environment.
+    """
+    def _send_custom_key(keystr):
+        """
+        Send custom keyword to SLOF's user interface.
+        """
+        logging.info('Sending \"%s\" to slof terminal.' % keystr)
+        for key in keystr:
+            if key == '-':
+                vm.send_key('minus')
+            else:
+                vm.send_key(key)
+        vm.send_key('ret')
+
+    def _create_iso():
+        """
+        Create iso with one file on it
+        """
+        if params.get("create_cd", 'no') == 'yes':
+            logging.info("creating iso.")
+            iso_name = params.get("cdrom_cd1")
+            iso = utils_misc.get_path(data_dir.get_data_dir(), iso_name)
+            process.system_output("dd if=/dev/urandom of=new bs=10M count=1",
+                                  300)
+            process.system_output("mkisofs -o %s new" % iso, 300)
+            process.system_output("rm -f new", 300)
+
+    def _cleanup_iso():
+        """
+        Removes created cdrom
+        """
+        if params.get("create_cd", 'no') == 'yes':
+            logging.info("cleaning up temp iso")
+            cdrom_test = utils_misc.get_path(
+                data_dir.get_data_dir(), params.get("cdrom_cd1"))
+            os.remove(cdrom_test)
+
+    def _verify_boot_once(content):
+        """
+        Verify the specify device if booted once.
+        """
+        parent_bus_type = 'pci'
+        if params['boot_once'] == 'n':
+            child_bus_type = params['net_type']
+            child_addr = params['nic_pci_addr_nic1']
+            if params['net_type'] == 'l-lan':
+                parent_bus_type = 'vdevice'
+                child_addr = params['spapr_vlan_reg']
+        elif params['boot_once'] == 'c':
+            child_bus_type = 'scsi'
+            child_addr = params['drive_pci_addr_empty_image1']
+        elif params['boot_once'] == 'd':
+            child_bus_type = 'scsi'
+            child_addr = params['bus_extra_params_cd1']
+        if slof.verify_boot_device(
+                content, parent_bus_type, child_bus_type, child_addr):
+            logging.info('Booted from %s@%s.' % (child_bus_type, child_addr))
+        else:
+            test.fail('Failed to boot from %s@%s once.' % (child_bus_type,
+                                                           child_addr))
+
+    def _verify_boot_order(content):
+        """
+        Verify the devices if booted by order.
+        """
+        Dev = namedtuple('Dev', 'parent, child, addr')
```

Comment:
why not define this `Dev` inside `slof.py`, thus `verify_boot_device` could be defined as:
```python
def verify_boot_device(content, dev,  sub_child_addr=None, position=0):
```

Reply:
My bad, this is not for slof_boot case, have updated.